protoc_plugin/test/proto3_json_test.dart

```diff
@@ -453,6 +453,81 @@ void main() {
             }, supportNamesWithUnderscores: false),
           parseFailure(['optional_foreign_message']));
     });
+    test('permissive enums', () {
+      final sparseB = SparseEnumMessage()..sparseEnum = TestSparseEnum.SPARSE_B;
+      expect(
+          SparseEnumMessage()..mergeFromProto3Json({'sparseEnum': 'SPARSE_B'}),
+          sparseB);
+      expect(
+          () => SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sparse_b'}),
+          parseFailure(['sparseEnum']));
+      expect(
+          () => SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'SPARSE-B'}),
+          parseFailure(['sparseEnum']));
+      expect(
+          () => SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sPaRsE_b'}),
+          parseFailure(['sparseEnum']));
+      expect(
+          () => SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sparseB'}),
+          parseFailure(['sparseEnum']));
+      expect(
+          () => SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'spaRSEB'}),
+          parseFailure(['sparseEnum']));
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sparse_b'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'SPARSE-B'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'S-P-A-R-S-E-B'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sPaRsE_b'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'sparseB'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          SparseEnumMessage()
+            ..mergeFromProto3Json({'sparseEnum': 'spaRSEB'},
+                permissiveEnums: true),
+          sparseB);
+      expect(
+          () => Any()
+            ..mergeFromProto3Json({
+              '@type':
+                  'type.googleapis.com/protobuf_unittest.SparseEnumMessage',
```

Comment:
Add tests with `-` or `_` at the start or end of the name.

Reply:
Done